gson/src/main/resources/META-INF/proguard/gson.pro

```diff
@@ -59,13 +54,16 @@
   <init>();
 }
 
-# If a class is used in some way by the application and has fields annotated with @SerializedName,
-# keep those fields and the constructors of the class
-# For convenience this also matches classes without no-args constructor, in which case Gson uses JDK Unsafe
-# Based on https://issuetracker.google.com/issues/150189783#comment11
-# See also https://github.com/google/gson/pull/2420#discussion_r1241813541 for a more detailed explanation
+# Keep fields annotated with @SerializedName for classes which are present.
+# If classes with fields annotated with @SerializedName have a no-args
+# constructor keep that as well.
 -if class *
--keepclasseswithmembers,allowobfuscation,allowoptimization class <1> {
-  <init>(...);
```

Comment:
Is it intentional that you removed `<init>(...)` again and used `<init>()` below? My personal opinion in https://github.com/google/gson/pull/2448#pullrequestreview-1553268244 was not blocking the merge of this pull request (sorry if I did not make that clear enough).

This change to use `<init>()` again might render some or all of the shrinker test changes obsolete.

Reply:
Or at least I am a bit surprised that no exception occurs for `ClassWithoutDefaultConstructor`; I assume it might be related to your newly added `-keepclasseswithmembers`.

So, do I understand it correctly that your new approach of using `-keepclasseswithmembers` achieves the same thing as your initial `<init>(...)` approach (i.e. Gson can instantiate classes without no-args constructor using Unsafe; instead of failing because R8 made the class abstract). But this new approach is better than the `<init>(...)` rule because that would have unnecessarily kept the constructors with args around, even though they are not actually used by Gson?